metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -362,50 +394,114 @@ final class FileDecoderProvider(
       sourceFile: AbsolutePath
   ): Either[
     String,
-    (BuildTargetIdentifier, ScalaTarget, AbsolutePath)
+    (BuildTargetIdentifier, ScalaTarget, AbsolutePath, AbsolutePath)
   ] = {
     val metadata = for {
       targetId <- buildTargets.inverseSources(sourceFile)
       target <- buildTargets.scalaTarget(targetId)
-      sourceRoot <- buildTargets.workspaceDirectory(targetId)
-    } yield (targetId, target, sourceRoot)
+      workspaceDirectory <- buildTargets.workspaceDirectory(targetId)
+      sourceRoot <- buildTargets.inverseSourceItem(sourceFile)
+    } yield (targetId, target, workspaceDirectory, sourceRoot)
     metadata.toRight(
       s"Cannot find build's metadata for ${sourceFile.toURI.toString()}"
     )
   }
 
   private def decodeJavapFromClassFile(
-      path: AbsolutePath,
       verbose: Boolean
-  ): Future[DecoderResponse] = {
+  )(path: AbsolutePath): Future[DecoderResponse] = {
     try {
       val args = if (verbose) List("-verbose") else Nil
-      val sb = new StringBuilder()
+      val sbOut = new StringBuilder()
+      val sbErr = new StringBuilder()
       shellRunner
         .run(
           "Decode using javap",
           JavaBinary(userConfig().javaHome, "javap") :: args ::: List(
             path.filename
           ),
           path.parent,
-          redirectErrorOutput = true,
+          redirectErrorOutput = false,
           Map.empty,
           s => {
-            sb.append(s)
-            sb.append(Properties.lineSeparator)
+            sbOut.append(s)
+            sbOut.append(Properties.lineSeparator)
+          },
+          s => {
+            sbErr.append(s)
+            sbErr.append(Properties.lineSeparator)
           },
-          s => (),
           propagateError = true,
           logInfo = false
         )
-        .map(_ => DecoderResponse.success(path.toURI, sb.toString()))
+        .map(_ => {
+          if (sbErr.nonEmpty)
+            DecoderResponse.failed(path.toURI, sbErr.toString)
+          else
+            DecoderResponse.success(path.toURI, sbOut.toString)
+        })
     } catch {
       case NonFatal(e) =>
         scribe.error(e.toString())
         Future.successful(DecoderResponse.failed(path.toURI, e))
     }
   }
 
+  private def decodeCFRFromClassFile(
```

Comment:
Could we run CFR using ShellRunner and download the dependency on demand? This would mean we wouldn't need to add the additional dependency to Metals.

I see there is a main that can be used https://github.com/leibnitz27/cfr/blob/9f5a97cf76d94d3ac695887abc6bbfb740c70a9c/src/org/benf/cfr/reader/Main.java

Reply:
Could do.  It's only 2MB but I guess it all adds up.  Is there someplace where this is already done - where I can nick the code?

Reply:
Yeah, it all adds up and especially if it's a tool used by more advanced developers this might not be needed to download at all 

That should be for example in the NewProjectProvider, it uses gitter8 by downloading it and running with the current java home.

Reply:
Here to be specific: 
 https://github.com/scalameta/metals/blob/main/metals/src/main/scala/scala/meta/internal/builds/NewProjectProvider.scala#L107

Reply:
Done - `ShellRunner#runJava` made this very easy to change.